Web Inspector: Use strict mode for String.prototype.* methods

stwrt opened the pull request:
#### 7983d66689670efba91a29632c2ef66df9154a89
<pre>
Web Inspector: Use strict mode for String.prototype.* methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=176074">https://bugs.webkit.org/show_bug.cgi?id=176074</a>
<a href="https://rdar.apple.com/problem/166488067">rdar://problem/166488067</a>

Reviewed by Devin Rousso.

Added &quot;use strict&quot; to missing String.prototype.* methods.
Also added tests for previously untested methods and tests verifying that
methods return primitive strings.

* LayoutTests/inspector/unit-tests/string-utilities-expected.txt:
* LayoutTests/inspector/unit-tests/string-utilities.html:
* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(get return):
(value):

Canonical link: <a href="https://commits.webkit.org/304532@main">https://commits.webkit.org/304532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad2eb44f28e82d9672b8be36ff8fe4c8b0b14ebe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135849 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47141 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143562 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7ae66253-4f50-4986-bed3-38f4c7c6d942) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8883 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8071 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103819 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ebfd7506-2bdc-41c5-99dd-5c2bb3882d6f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121755 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84696 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5a2b9140-b850-408c-98ce-842b2a686eee) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6132 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3773 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4166 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39950 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146310 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7910 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40510 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112176 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7937 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6640 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112562 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28569 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6039 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118055 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61817 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7957 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36138 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7691 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71508 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7912 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7774 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso reviewed (commented):
so i love the tests but im not really sure what benefit adding `"use strict";` provides in these functions

is there a reason for doing this?  the original bug claims that without it the functions are slower.  is that actually still true?

if so, why not apply this to _all_ free functions (i.e. not in a `class`)?

dcrousso commented on LayoutTests/inspector/unit-tests/string-utilities.html:
```diff
@@ -161,6 +234,41 @@
+            InspectorTest.expectEqual(typeof "abc".truncate(10), "string", "truncate should return a primitive string.");
+
+            // Verify equality comparisons work correctly (would fail if String objects were returned).
+            InspectorTest.expectThat("abc".collapseWhitespace() === "abc", "collapseWhitespace result should be strictly equal to primitive string.");
```
let's use `expectEqual` so that if this fails we get a better error message

ditto below

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1255,6 +1279,8 @@ Object.defineProperty(String.prototype, "levenshteinDistance",
 {
     value(s)
```
`String.prototype.levenshteinDistance` actually doesnt appear to be used anywhere so we should just remove it instead

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1288,6 +1314,8 @@ Object.defineProperty(String.prototype, "toCamelCase",
 {
     value()
     {
+        "use strict";
```
ditto with removing this

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1296,6 +1324,8 @@ Object.defineProperty(String.prototype, "hasMatchingEscapedQuotes",
 {
     value()
     {
+        "use strict";
```
ditto with removing this

dcrousso reviewed (approved):
neat!

stwrt committed:
Web Inspector: Use strict mode for String.prototype.* methods
https://bugs.webkit.org/show_bug.cgi?id=176074
rdar://problem/166488067

Reviewed by Devin Rousso.

Added "use strict" to missing String.prototype.* methods.
Also added tests for previously untested methods and tests verifying that
methods return primitive strings.

* LayoutTests/inspector/unit-tests/string-utilities-expected.txt:
* LayoutTests/inspector/unit-tests/string-utilities.html:
* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(get return):
(value):

Canonical link: https://commits.webkit.org/304532@main
